src/renderers/shared/fiber/ReactFiberScheduler.js

```diff
@@ -454,162 +445,209 @@ module.exports = function<T, P, I, TI, C>(config : HostConfig<T, P, I, TI, C>) {
+      } else {
+        return;
       }
-    } finally {
-      isPerformingTaskWork = false;
     }
   }
 
-  function performTaskWork() {
-    performAndHandleErrors(TaskPriority);
-  }
+  function performWork(priorityLevel : PriorityLevel, deadline : null | Deadline) {
+    if (isPerformingWork) {
+      throw new Error('performWork was called recursively.');
+    }
+    isPerformingWork = true;
 
-  function performAndHandleErrors(priorityLevel : PriorityLevel, deadline : null | Deadline) {
-    // The exact priority level doesn't matter, so long as it's in range of the
-    // work (sync, animation, deferred) being performed.
-    let shouldContinue = true;
-    while (shouldContinue) {
-      shouldContinue = false;
-      const prevShouldBatchUpdates = shouldBatchUpdates;
-      shouldBatchUpdates = true;
+    // Perform work until either there's no more work at this priority level, or
+    // (in the case of deferred work) we've run out of time.
+    while (priorityLevel !== NoWork) {
+      // Functions that contain a try-catch block are not optimizable by the
+      // JS engine. The hottest code paths have been extracted to separate
+      // functions, workLoop and deferredWorkLoop, which run on every unit of
+      // work. The loop we're in now runs infrequently: to flush task work at
+      // the end of a frame, or to restart after an error.
       try {
-        switch (priorityLevel) {
+        if (priorityLevel >= HighPriority) {
+          if (!deadline) {
+            throw new Error('Cannot perform deferred work without a deadline.');
+          }
+          // The deferred work loop will run until there's no time left in
+          // the current frame
+          deferredWorkLoop(deadline);
+        } else {
+          // The non-deferred work loop will run until there's no more work
+          // at the given priority level
+          workLoop(priorityLevel);
+        }
+
+        // Stop performing work
+        priorityLevel = NoWork;
+
+        // There might still be work left. Depending on the priority, we should
+        // either perform it now or schedule a callback to perform it later.
+        switch (nextPriorityLevel) {
           case SynchronousPriority:
-            performSynchronousWorkUnsafe();
-            break;
           case TaskPriority:
-            if (!isPerformingTaskWork) {
-              performTaskWorkUnsafe();
-            }
+            // Perform work immediately by switching the priority level
+            // and continuing the loop.
+            priorityLevel = nextPriorityLevel;
             break;
           case AnimationPriority:
-            performAnimationWorkUnsafe();
+            scheduleAnimationCallback(performAnimationWork);
+            // Even though the next unit of work has animation priority, there
+            // may still be deferred work left over as well. I think this is
+            // only important for unit tests. In a real app, a deferred callback
```

Comment:
Yeah let's do this as a follow-up. I like the `flushWithCycles` idea. The way `flushDeferredPri` works right now is confusing because it takes 5(n+1) to do n cycles of work.